src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,712 @@
+    permuted_table_value: Polynomial<C::Scalar, LagrangeCoeff>,
+    permuted_table_poly: Polynomial<C::Scalar, Coeff>,
+    permuted_table_coset: Polynomial<C::Scalar, ExtendedLagrangeCoeff>,
+    permuted_table_blind: Blind<C::Scalar>,
+    permuted_table_commitment: C,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Product<C: CurveAffine> {
+    product_poly: Polynomial<C::Scalar, Coeff>,
+    product_coset: Polynomial<C::Scalar, ExtendedLagrangeCoeff>,
+    product_inv_coset: Polynomial<C::Scalar, ExtendedLagrangeCoeff>,
+    product_blind: Blind<C::Scalar>,
+    product_commitment: C,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Committed<C: CurveAffine> {
+    pub permuted: Permuted<C>,
+    pub product: Product<C>,
+}
+
+pub(crate) struct Constructed<C: CurveAffine> {
+    permuted_input_poly: Polynomial<C::Scalar, Coeff>,
+    permuted_input_blind: Blind<C::Scalar>,
+    permuted_input_commitment: C,
+    permuted_table_poly: Polynomial<C::Scalar, Coeff>,
+    permuted_table_blind: Blind<C::Scalar>,
+    permuted_table_commitment: C,
+    product_poly: Polynomial<C::Scalar, Coeff>,
+    product_blind: Blind<C::Scalar>,
+    product_commitment: C,
+}
+
+pub(crate) struct Evaluated<C: CurveAffine> {
+    constructed: Constructed<C>,
+    pub product_eval: C::Scalar,
+    pub product_inv_eval: C::Scalar,
+    pub permuted_input_eval: C::Scalar,
+    pub permuted_input_inv_eval: C::Scalar,
+    pub permuted_table_eval: C::Scalar,
+}
+
+impl Argument {
+    /// Given a Lookup with input columns [A_0, A_1, ..., A_m] and table columns
+    /// [S_0, S_1, ..., S_m], this method
+    /// - constructs A_compressed = A_0 + theta A_1 + theta^2 A_2 + ... and
+    ///   S_compressed = S_0 + theta S_1 + theta^2 S_2 + ...,
+    /// - permutes A_compressed and S_compressed using permute_column_pair() helper,
+    ///   obtaining A' and S', and
+    /// - constructs Permuted<C> struct using permuted_input_value = A', and
+    ///   permuted_table_value = S'.
+    /// The Permuted<C> struct is used to update the Lookup, and is then returned.
+    pub(in crate::plonk) fn commit_permuted<
+        C: CurveAffine,
+        HBase: Hasher<C::Base>,
+        HScalar: Hasher<C::Scalar>,
+    >(
+        &self,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        domain: &EvaluationDomain<C::Scalar>,
+        theta: ChallengeTheta<C::Scalar>,
+        advice_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        fixed_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        aux_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        transcript: &mut Transcript<C, HBase, HScalar>,
+    ) -> Result<Permuted<C>, Error> {
+        // Values of input columns involved in the lookup
+        let unpermuted_input_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>> = self
+            .input_columns
+            .iter()
+            .map(|&input| match input.column_type() {
+                Any::Advice => advice_values[input.index()].clone(),
```

Comment:
Instead of cloning these, obtain references to them (we're not mutating them). Further, place them into `Permuted<C>` so that when you call `commit_product` you don't need to retrieve them again -- which will then allow you to define `commit_product` on `Permuted<C>` rather than on `Argument`. (This will make it so that `Permuted<C>` will need a lifetime bound added to it.)